src/core/experimental/rcptr.d

```diff
@@ -0,0 +1,208 @@
+
+    __rcptr!int a; //default ctor
+    assert(a.get is null);
+
+    auto b = __rcptr!int(allocInts(10)); //ptr ctor
+
+    {
+        auto c = a; //copy ctor
+        assert(c.get == a.get);
+        assert(c.get != b.get);
+
+        c = b; //opAssign
+        assert(c.get != a.get);
+        assert(c.get == b.get);
+    }
+
+    assert(a.count is null);
+    assert(*b.count == 0);
+}
+
+/**
+A reference counted shared pointer that can be used to implement reference
+counted data structures.
+
+The reference count is automatically incremented/decremented on assignment,
+copy (construction), and destruction. When there are no more references to the
+pointer, the reference count is automatically deallocated and the referenced
+pointer is `free`d.
+
+Implementation: The internal implementation of `__rcptr` uses `malloc`/`free`.
+*/
+struct __rcptr(T)
+{
+    alias CounterType = uint;
+
+    private T* ptr = null;
+    private shared(CounterType)* count = null;
+
+    /**
+    Creates a new `__rcptr` instance, tracking the provided pointer.
+
+    This implies that the ownership of the pointer is transferred to
+    `__rcptr`.
+
+    Params:
+         ptr = pointer to memory to be managed by `__rcptr`
+    */
+    this(inout(T*) ptr) inout
+    {
+        this.ptr = ptr;
+
+        if (ptr !is null)
+        {
+            import core.memory : pureCalloc;
+
+            // We use `calloc` so we don't have to manually initialise count/addRef
+            count = (() @trusted inout => cast(typeof(count)) pureCalloc(1, CounterType.sizeof))();
+        }
+    }
+
+    /**
+    Get the owned pointer.
+
+    Returns:
+        The pointer owned by this instance of `__rcptr`.
+    */
+    @system
+    inout(T*) get() inout
+    {
+        return ptr;
+    }
+
+    ~this()
+    {
+        delRef();
+    }
+
+    ///
+    void opAssign(ref __rcptr!T rhs)
+    {
+        if (rhs.count == count)
+        {
+            return;
+        }
+
+        delRef();
+
+        ptr = rhs.ptr;
+        count = rhs.count;
+
+        addRef();
+    }
+
+    ///
+    this(scope ref inout __rcptr!T rhs) inout
+    {
+        ptr = rhs.ptr;
+        count = rhs.count;
+
+        addRef();
+    }
+
+    private void addRef() inout
+    {
+        import core.atomic : atomicOp;
+
+        if (ptr is null)
+        {
+            return;
+        }
+
+        () @trusted { atomicOp!"+="(*(cast(shared(CounterType)*) count), 1); } ();
+    }
+
+    private void delRef()
+    {
+        import core.atomic : atomicOp;
+
+        if (ptr is null)
+        {
+            return;
+        }
+
+        // The counter is left at -1 when this was the last reference
+        // (i.e. the counter is 0-based, because we use calloc)
+        if (atomicOp!"-="(*count, 1) == -1)
+        {
+            deallocate();
+        }
+    }
+
+    @trusted
+    private void deallocate()
+    {
+        import core.memory : pureFree;
+
+        pureFree(ptr);
```

Comment:
So this is a big problem here as we assume that `ptr` as allocated via `malloc`. Maybe we can templatize `__rcptr` and allow the user to provide his own deallocate function? Alternatively, we would need to store a pointer to such a deallocation function in the `count` payload.

Reply:
If we modify `__rcptr` to use intrusive ref counting we could properly support custom allocators (once we have them in druntime).

For now, should I just add a template parameter?

Reply:
Sadly, I'm not sure how we'll be able to use custom allocators as we'll be hit by the same problem as in the past: calling allocate / deallocate on an immutable allocator instance.

The very first implementation that I tried (long ago) was with intrusive reference counting and custom allocators and that hit a lot of bumps down the road. This is why we've over-simplified the design, going from an array with intrusive RC and custom allocators support to the implementation of `rcarray`

Reply:
> This is why we've over-simplified the design, going from an array with intrusive RC and custom allocators support to the implementation of rcarray

I understand the troubles you face, but this solution kinda buys us nothing. At least, it buys me nothing personally.
I won't use not-invasive ref counting.

What are we trying to do with this solution? Is it a stop-gap, or is it a thing?

Reply:
> I won't use not-invasive ref counting.

Why not?

> What are we trying to do with this solution?

Library ref-counting for RefCounting for rcarray & co . with some (future) compiler magic.

Reply:
It would be incompatible with any of my existing code. It's also inefficient.

What compiler magic are we envisioning?

Reply:
> It would be incompatible with any of my existing code. 

This isn't intended to be for users. Hence, __

> What compiler magic are we envisioning?

I think it was planned to detect __RefCounted for better @safe safety, but it currently looks like Water is trying to generalize this with DIP 1021.

Reply:
> > It would be incompatible with any of my existing code.
> 
> This isn't intended to be for users. Hence, __

I'm not sure what you mean... how do I make use of it then? What does it offer if not a platform the specifying how ref-counting should function?

I want to know why `opInc`/`opDec` failed? After 10 years of waiting for this, this direction is embarrassingly feeble, and even slightly worse than C++, which we determined was so disappointing that we spent **TEN YEARS** trying to do something better. If we can't do something *at least* as good as C++, then common-sense suggests we just copy C++, at least then we'll be fully compatible with C++.

Reply:
> how do I make use of it then?

You don't.


>  What does it offer if not a platform specifying how ref-counting should function?

It's internal for now and the forseeable future as no one is fully sure on the direction.

> I want to know why opInc/opDec failed? 

I don't know.

> this direction is embarrassingly feeble

We need rcarray __now__, but we don't want to expose the exact details of it yet. What's so wrong with that?

BTW I'm the wrong person to ask these questions to.

Reply:
> > how do I make use of it then?
> 
> You don't.

Huh? Why are we introducing code that has no use?

> > What does it offer if not a platform specifying how ref-counting should function?
> 
> It's internal for now and the forseeable future as no one is fully sure on the direction.

Evidently.

> > I want to know why opInc/opDec failed?
> 
> I don't know.

I feel like we need to understand this very clearly.

> > this direction is embarrassingly feeble
> 
> We need rcarray **now**, but we don't want to expose the exact details of it yet. What's so wrong with that?

What do we need it for? We've done without it for a very long time...

> BTW I'm the wrong person to ask these questions to.

Okay, but you're the only one posting on the thread that appears to have a sense of authority :)

Reply:
> Huh? Why are we introducing code that has no use?

It has a used. It will be immediately used for rcarray, but allowing us to not commit to the interface/design just yet.

> What do we need it for? We've done without it for a very long time... 

Sure, and I survived without a million dollars for a very long time too. However, we need to start somewhere if the vision of a fully @nogc D is ever to become a reality.

> you're the only one posting on the thread that appears to have a sense of authority

Yeah that's pretty sad. Maybe this conversation is better moved to the NG?